File: deploy/claude/guards.py
"""The context guard of a place, as the panel's executor keeps it for the host."""

import os

CAP_DEFAULT = 80


def path():
    """Returns the file the executor keeps the guards in."""
    base = os.environ.get("XDG_STATE_HOME") or os.path.join(os.path.expanduser("~"), ".local", "state")
    return os.path.join(base, "aacpanel", "guards.tsv")


def where(payload):
    """Returns the directory a session belongs to.

    The project directory claude names for its hooks, not the working directory:
    that one follows every cd the agent makes, and the cap would jump to whatever
    project the agent is looking into.
    """
    return (os.environ.get("CLAUDE_PROJECT_DIR")
            or (payload.get("cwd") if isinstance(payload, dict) else "")
            or os.getcwd())


def of(cwd, file=None):
    """Returns (cap, restart) of the place closest above cwd, or None where the file names none.

    A line a place: the directory, the cap in percent of the model's window, and 1
    where a session past it restarts itself. The closest place is the longest path
    that is the directory itself or holds it — a project inside its contour's tree.
    """
    try:
        with open(file or path(), encoding="utf-8") as f:
            lines = f.read().splitlines()
    except OSError:
        return None
    best = None
    for line in lines:
        parts = line.split("\t")
        if len(parts) != 3:
            continue
        place, cap, restart = parts
        if not cwd or not (cwd == place or cwd.startswith(place.rstrip("/") + "/")):
            continue
        if best is not None and len(place) <= len(best[0]):
            continue
        try:
            best = (place, int(cap), restart == "1")
        except ValueError:
            continue
    return None if best is None else best[1:]

File: internal/executor/guards.go
package executor

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"slices"
	"strconv"
	"strings"

	"aacpanel/internal/action"
)

// GuardsPath is where the host reads the context guard of every place the map
// knows: the context guard hook and the prompt stamp of every contour, run as
// this user. A line a place — the directory, the cap in percent of the model's
// window, and 1 where a session past it restarts itself — so that a hook in
// any language, a shell one included, reads it without a parser.
func GuardsPath() string {
	base := os.Getenv("XDG_STATE_HOME")
	if base == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			home = os.TempDir()
		}
		base = filepath.Join(home, ".local", "state")
	}
	return filepath.Join(base, "aacpanel", "guards.tsv")
}

// KeepGuards writes the guards the service computed from the map. The file is
// replaced whole by a rename: a hook reading it mid-write sees the old places
// or the new ones, never half of each.
func (e *Executor) KeepGuards(_ context.Context, guards []action.Guard) error {
	lines := make([]string, 0, len(guards))
	for _, g := range guards {
		restart := "0"
		if g.Restart {
			restart = "1"
		}
		lines = append(lines, g.Path+"\t"+strconv.Itoa(g.Cap)+"\t"+restart)
	}
	slices.Sort(lines)

	path := GuardsPath()
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return fmt.Errorf("the guards directory: %w", err)
	}
	tmp, err := os.CreateTemp(filepath.Dir(path), ".guards-*")
	if err != nil {
		return fmt.Errorf("the guards file: %w", err)
	}
	defer os.Remove(tmp.Name())
	body := strings.Join(lines, "\n")
	if body != "" {
		body += "\n"
	}
	if _, err := tmp.WriteString(body); err != nil {
		tmp.Close()
		return fmt.Errorf("the guards file: %w", err)
	}
	if err := tmp.Close(); err != nil {
		return fmt.Errorf("the guards file: %w", err)
	}
	if err := os.Chmod(tmp.Name(), 0o600); err != nil {
		return fmt.Errorf("the guards file: %w", err)
	}
	return os.Rename(tmp.Name(), path)
}
